From abca61fad742ce147f8877862f072a28df700e2b Mon Sep 17 00:00:00 2001
From: Robert Norton <ronorton@microsoft.com>
Date: Fri, 3 Feb 2023 13:13:17 +0000
Subject: [PATCH 09/10] Add a hacky mechanism to track new lines in terminal
 and trace output and insert extra new lines to keep them separate.

---
 c_emulator/riscv_platform_impl.c |  6 ++++++
 c_emulator/riscv_prelude.c       | 20 ++++++++++++++++++++
 2 files changed, 26 insertions(+)

diff --git a/c_emulator/riscv_platform_impl.c b/c_emulator/riscv_platform_impl.c
index 9aca0b7..e554d09 100644
--- a/c_emulator/riscv_platform_impl.c
+++ b/c_emulator/riscv_platform_impl.c
@@ -44,9 +44,15 @@ uint64_t rv_htif_tohost = UINT64_C(0x80001000);
 uint64_t rv_insns_per_tick = UINT64_C(100);
 
 int term_fd = 1; // set during startup
+extern bool have_newline;
 void plat_term_write_impl(char c)
 {
   if (write(term_fd, &c, sizeof(c)) < 0) {
     fprintf(stderr, "Unable to write to terminal!\n");
   }
+  /* Trace output always goes to stdout. If the terminal is also going to stdout
+   * we try to separate them by inserting new lines before trace output when
+   * necessary (see riscv_prelude.c). */
+  if (term_fd == 1)
+    have_newline = c == '\n';
 }
diff --git a/c_emulator/riscv_prelude.c b/c_emulator/riscv_prelude.c
index ad5a831..2235ba0 100644
--- a/c_emulator/riscv_prelude.c
+++ b/c_emulator/riscv_prelude.c
@@ -1,33 +1,53 @@
 #include "riscv_prelude.h"
 #include "riscv_config.h"
 
+/* Attempt to keep track of whether output is at the start of a line so that we
+ * can produce nicer output if trace output interrupts terminal output. */
+bool have_newline = true;
+
+void ensure_newline()
+{
+  if (!have_newline)
+    putchar('\n');
+}
+
 unit print_string(sail_string prefix, sail_string msg)
 {
+  ensure_newline();
   printf("%s%s\n", prefix, msg);
+  have_newline = true;
   return UNIT;
 }
 
 unit print_instr(sail_string s)
 {
+  ensure_newline();
   if (config_print_instr) printf("%s\n", s);
+  have_newline = true;
   return UNIT;
 }
 
 unit print_reg(sail_string s)
 {
+  ensure_newline();
   if (config_print_reg) printf("%s\n", s);
+  have_newline = true;
   return UNIT;
 }
 
 unit print_mem_access(sail_string s)
 {
+  ensure_newline();
   if (config_print_mem_access) printf("%s\n", s);
+  have_newline = true;
   return UNIT;
 }
 
 unit print_platform(sail_string s)
 {
+  ensure_newline();
   if (config_print_platform) printf("%s\n", s);
+  have_newline = true;
   return UNIT;
 }
 
-- 
2.39.2

